nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/test/java/org/apache/nifi/processors/email/TestListenSMTP.java

```diff
@@ -13,307 +13,174 @@
-            splits.get(0).assertAttributeEquals("smtp.to", "bob@nifi.apache.org");
-
-            Thread.sleep(100);
-        } finally {
-            // shut down the server
-            listenSmtp.startShutdown();
-        }
+    /**
+     *
+     */
+    @After
+    public void after() {
+        this.executor.shutdown();
     }
 
-    @Test(timeout=15000)
-    public void ValidEmail() throws Exception, EmailException {
-        final boolean[] failed = {false};
-        ListenSMTP listenSmtp = new ListenSMTP();
-        final TestRunner runner = TestRunners.newTestRunner(listenSmtp);
+    /**
+     *
+     */
+    @Test
+    public void validateSuccessfulInteraction() throws Exception, EmailException {
+        int port = NetworkUtils.availablePort();
 
-        runner.setProperty(ListenSMTP.SMTP_PORT, "0");
-        runner.setProperty(ListenSMTP.SMTP_HOSTNAME, "bermudatriangle");
+        TestRunner runner = TestRunners.newTestRunner(ListenSMTP.class);
+        runner.setProperty(ListenSMTP.SMTP_PORT, String.valueOf(port));
         runner.setProperty(ListenSMTP.SMTP_MAXIMUM_CONNECTIONS, "3");
         runner.setProperty(ListenSMTP.SMTP_TIMEOUT, "10 seconds");
 
-        final ProcessSessionFactory processSessionFactory = runner.getProcessSessionFactory();
-        final ProcessContext context = runner.getProcessContext();
+        runner.assertValid();
 
-        // NOTE: This test routine uses  the same strategy used by TestListenAndPutSyslog
-        // where listenSmtp method calls are used to allow the processor to be started using
-        // port "0" without triggering a violation of PORT_VALIDATOR
-        listenSmtp.onScheduled(context);
-        listenSmtp.initializeSMTPServer(context);
+        int messageCount = 5;
+        CountDownLatch latch = new CountDownLatch(messageCount);
 
-        final int port = listenSmtp.getPort();
+        this.executor.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                runner.run(1, false);
+            }
+        }, 0, 500, TimeUnit.MILLISECONDS);
 
-        try {
-            final Thread clientThread = new Thread(new Runnable() {
-                @Override
-                public void run() {
+        this.executor.schedule(new Runnable() {
+            @Override
+            public void run() {
+                for (int i = 0; i < messageCount; i++) {
                     try {
                         Email email = new SimpleEmail();
-                        email.setHostName("127.0.0.1");
+                        email.setHostName("localhost");
                         email.setSmtpPort(port);
-                        email.setStartTLSEnabled(false);
                         email.setFrom("alice@nifi.apache.org");
                         email.setSubject("This is a test");
-                        email.setMsg("Test test test chocolate");
+                        email.setMsg("MSG-" + i);
```

Comment:
/joke on
NOOOOOOOOOO!!!! What have you done!??!! :-)

https://github.com/apache/nifi/search?utf8=%E2%9C%93&q=chocolate
/joke off